Update 12-Testing_for_Command_Injection.md (#1243)
OWASP/wstg: document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/12-Testing_for_Command_Injection.md

This article describes how to test an application for OS command injection. The tester will try to inject an OS command through an HTTP request to the application.

> OS commmand injection is a vulnerability that occurs when user input is directly passed to an operating system command without proper validation or sanitization. This allows the user to inject and execute arbitrary commands on the server which can lead to unauthorized data access, data corruption and full server compromise. This vulnerability can be prevented by emphasizing security during the design and development of applications.

Reviewer: ```suggestion
OS commmand injection is a vulnerability that occurs when user input is directly passed to an operating system command without proper validation or sanitization. This allows the user to inject and execute arbitrary commands on the server which can lead to unauthorized data access, data corruption, and full server compromise. This vulnerability can be prevented by emphasizing security during the design and development of applications.
```

For instance, the following special characters can be used on both Windows and Unix-based systems :
- `cmd1|cmd2` : cmd2 will be executed whether cmd1 succeeds or not.
- `cmd1||cmd2` : cmd2 will only be executed if cmd1 fails.
- `cmd1&&cmd2` : cmd2 will only be executed if cmd1 succeeds.
- `cmd1&cmd2` : cmd2 will be executed whether cmd1 succeeds or not.  

Note that, `;` will work on Unix-based systems and PowerShell. However, it will not work on Windows Command Prompt (cmd).  
Furthermore, you can use Bash command substitution `$(cmd)` or ``cmd`` to execute commands on Unix-based systems.  
Additionally, Linux file descriptors such as `>(cmd)`, `<(cmd)` can also be used.

## Filter Evasion
To prevent OS command injection, web developers often use filters. However, these filters are sometimes not properly implemented which allows attackers to bypass them.
In this section, we will cover different techniques used to bypass those filters.

### Methodology
First of all, it is always good practice to have a basic understanding of how the filter works before trying to bypass it.
Here is a methodology we can use when we come across a filter:
- Is the filter client-side or server-side ?

> - Is the filter applied on special characters, os commands, or both ?

Author: Done

#### Bash Brace Expansion
Bash brace expansion is a Bash feature that allows you to execute commands by using curly braces.  
For example, `{ls,-la}` will execute `ls -la` command. This can be extremely useful if the web server is filtering space, new line or tab characters.  
Let's assume that we want to display the content of '/etc/passwd' file. Thus, instead of using `;cat /etc/passwd`, we can use `;{cat,/etc/passwd}`  
That said, it's important to note that this technique will only work if the web server is using Bash and if characters like `}{/,;` are not filtered.

### Commands Filter Evasion
In this section, we are going to explore some techniques used to bypass filters applied on operating system commands.
#### Base64 Encoding
In certain scenarios, the web server may filter commands such as `whoami`, `id`, etc.  
Let's suppose that `whoami` is blocked by the web server. Therefore, we cannot use a payload like `;whoami`.  
To bypass this restriction, we will use instead the base64 encoded format of `whoami` by executing: `echo -n 'whoami' | base64`. This will return `d2hvYW1p`  

> After that, we will send the following payload: `;bash<<<$(base64 -d<<< d2hvYW1p)` to the vulnerable parameter. This should then bypass the server-side filter and allow us to achieve remote code execution.  

Reviewer: ```suggestion
After that, we will send the following payload: `;bash<<<$(base64 -d<<< d2hvYW1p)` in the vulnerable parameter. This should then bypass the server-side filter and allow us to achieve remote command execution.  
```

#### Case Modification
Changing the case of a command may allow us to bypass OS command filters.  

> Note that this will generally work on Windows systems see that commands are case sensitive in Linux.  

Reviewer: Seems like something is missing in this sentence. Also it should be "...on Linux."

Author: I made few changes. Let me know what you think

For instance, if we notice that the web server is blocking `;whoami`, we can try to use `;WhoAmi`.  
To use uppercase commands in Linux, we can use a technique called **character shifting** that works as follows:  
`;$(tr "[A-Z]" "[a-z]"<<<"WhoaMi")`  

> The command above will simply translate each uppercase character to its corresponding lowercase character.

Reviewer: Probably makes more sense to put this before the URL and base64 examples.

#### Character Insertion

> Characters like `\`; `$@`, `'` can be inserted to Linux OS commands without affecting the normal execution of the command.  

Reviewer: to ➡️ into

> For example, `who\ami`, `w$@hoami` or `wh'o'ami` will all execute the `whoami` command   

Reviewer: ```suggestion
For example, `who\ami`, `w$@hoami`, or `wh'o'ami` will all execute the `whoami` command   
```

Note that the number of single quotes `'` in the Linux command must be **even**, otherwise you will get an error. If you're using Windows CMD, make sure to use double quotes `"` instead.  
Furthermore, you can also use a caret `^` in CMD commands. For example, `whoa^mi` will execute the `whoami` command. This does not work in PowerShell.  

**Finally, note that you may need to combine different special characters and OS command filter evasion techniques to successfully bypass the filters put it place by the web server.**

## Blind Command Injection

> Sometimes, we may not be able to see the output from our injected command in the web server's HTTP response. Thus, we will need to find a way to confirm whether or not our injection succeeded. To do that, we can use `HTTP`, `DNS` or `SMTP` remote servers under our control.  

Reviewer: ```suggestion
Sometimes, we may not be able to see the output from our injected command in the web server's HTTP response. Thus, we will need to find a way to confirm whether or not our injection succeeded. To do that, we can use `HTTP`, `DNS`, or `SMTP` remote servers under our control.  
```

> We can also use **time delay system commands** like `sleep` (Linux), `timeout` (Windows) or network utility like `ping`.  

Reviewer: ```suggestion
We can also use **time delay system commands** like `sleep` (Linux), `timeout` (Windows), or network utility like `ping`.  
```

- Is the webapp using a allowlist or blocklist filter ?
- What OS is running on the web server? This allows us to have an idea of the commands and special characters we can use.

### Special Characters Filter Evasion

As previously mentionned, filters can either be applied on special characters, OS commands or both.
To bypass filters applied on special characters, we can use environment variables, Bash brace expansion or URL encoding.

#### URL Encoding

URL encoding special characters can allow us to bypass the filter if the web server only blocks the plaintext special characters.
Here are some special characters with their URL encoded format:

|Special Character|URL Encoding|
|-----------------|------------|
|;                | %3b        |
|space            | %20        |
|tab              | %09        |
|&                | %26        |
|New line         | %0a        |

For instance, instead of using `;whoami`, we could use `%3bwhoami`.

#### Environment Variables

Special characters like space, semi-colon, tab, or new line will generally be filtered by the web server especially if they are not useful for the specified input.  

> To escape this restriction, we can use environment variables such as **IFS**, **PATH** or **LS_COLORS** in Linux and **HOMEPATH** in Windows.  

Reviewer: ```suggestion
To escape this restriction, we can use environment variables such as **IFS**, **PATH** or **LS_COLORS** on Linux, and **HOMEPATH** on Windows.  
```

> For instance, on Linux, `/`, `;` and `[space]` can be replaced respectively with `${PATH:0:1}`, `${LS_COLORS:10:1}`, and `${IFS}`.  

Reviewer: ```suggestion
For instance, on Linux, `/`, `;`, and `[space]` can be replaced respectively with `${PATH:0:1}`, `${LS_COLORS:10:1}`, and `${IFS}`.  
```

- `exec`
- `proc_open`
- `eval`
- `passthru`

## Remediation

### Sanitization

The URL's query parameters and form data need to be validated and sanitized to prevent the injection of malicious characters.  

> A blacklist of characters is an option but it may be difficult to think of all of the characters to validate against. Also there may be some that were not discovered as of yet.  

Reviewer: blocklist

> A whitelist containing only authorized characters or commands should be created to validate the user input. Characters that were missed, as well as undiscovered threats, should be eliminated by this list.  

Reviewer: An allowlist

- `exec`
- `proc_open`
- `eval`
- `passthru`

## Remediation

### Sanitization

The URL query parameters and form data need to be validated and sanitized to prevent the injection of malicious characters.  
A blocklist of characters is an option but it may be difficult to think of all of the characters to validate against. Also there may be some that were not discovered as of yet.  
A allowlist containing only authorized characters or commands should be created to validate the user input. Characters that were missed, as well as undiscovered threats, should be eliminated by this list.  

General deny list to be included for command injection can be `|` `;` `&` `$` `>` `<` `'` `\` `!` `>>` `#`  

Escape or filter special characters for windows, `(` `)` `<` `>` `&` `*` `‘` `|` `=` `?` `;` `[` `]` `^` `~` `!` `.` `"` `%` `@` `/` `\` `:` `+` `,`  ``` ` ```  

Escape or filter special characters for Linux, `{` `}` `(` `)` `>` `<` `&` `*` `‘` `|` `=` `?` `;` `[` `]` `$` `–` `#` `~` `!` `.` `"` `%`  `/` `\` `:` `+` `,` ``` ` ```  

> Moreover, avoid using functions that execute operating system commands in your application unless absolutely necessary. For instance, instead of using `system("cp /path/to/file1.txt /path/to/file2.txt")` to copy a file, you can directly use the php built-in function `copy("cp /path/to/file1.txt, /path/to/file2.txt")` which does exactly the same thing.

Reviewer: Add something before the last period like: 

> (other development languages/frameworks generally offer similar built-in functionality.)

or change more of the sentence

> ...you can directly use a language or framework built-in function; such as PHP's `copy("cp /path/to/file1.txt, /path/to/file2.txt")` which does exactly the same thing.
